Can't create an Application

1. Add custom runtime:
```
{
  "name": "hydrosphere/serving-grpc-runtime-spark-2_1",
  "version": "0.0.1",
  "modelTypes": [
    "spark:2.1"
  ],
  "tags": [
    "string"
  ],
  "configParams": {}
}
```
2. Build a model (word2vec for Spark 2.1.0)
```
{
  "modelId": 14
}
```
3. Build an Application:
```
{
  "id": 0,
  "name": "testapp",
  "executionGraph": {
    "stages": [
      {
        "services": [
          {
            "serviceDescription": {
              "runtimeId": 2,
              "modelVersionId": 1,
              "environmentId": 0
            },
            "weight": 100
          }
        ],
        "signatureName": "string"
      }
    ]
  }
}
```
Result:
```
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null,Step 1/6 : FROM busybox:1.28.0,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null,
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null, ---> 5b0d59026729
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null,Step 2/6 : LABEL MODEL_TYPE=spark:2.1,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null,
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null, ---> Running in b5f914e59880
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null,Removing intermediate container b5f914e59880
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null, ---> df1966d1a6e8
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null,Step 3/6 : LABEL MODEL_NAME=word2vec,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null,
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null, ---> Running in 2172c7a57456
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null,Removing intermediate container 2172c7a57456
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null, ---> 5d32ba9e6aff
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null,Step 4/6 : LABEL MODEL_VERSION=None,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null,
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null, ---> Running in 877af9e80f15
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null,Removing intermediate container 877af9e80f15
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null, ---> b007c4cec8f3
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null,Step 5/6 : VOLUME /model,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null,
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null, ---> Running in e430f440c583
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null,Removing intermediate container e430f440c583
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null, ---> ede9188f3e2a
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null,Step 6/6 : ADD model /model,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null,
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null, ---> 8739a8be79c8
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null,null,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null,Successfully built 8739a8be79c8
,null,null,None)
[INFO ] i.h.s.m.s.ModelManagementServiceImpl i.h.s.m.s.ModelManagementServiceImpl$$anon$1.handle.257 ProgressMessage(null,null,Successfully tagged word2vec:1
,null,null,None)
[INFO ] c.s.d.c.DefaultDockerClient c.s.d.c.DefaultDockerClient.createContainer.635 Creating container with ContainerConfig: ContainerConfig{hostname=null, domainname=null, user=null, attachStdin=null, attachStdout=null, attachStderr=null, portSpecs=null, exposedPorts=null, tty=null, openStdin=null, stdinOnce=null, env=null, cmd=null, image=word2vec:1, volumes={/model={}}, workingDir=null, entrypoint=null, networkDisabled=null, onBuild=null, labels={MODEL_VERSION_ID=1, SERVICE_ID=1, MODEL_NAME=word2vec, MODEL_TYPE=spark:2.1, MODEL_VERSION=1, HS_SERVICE_MARKER=HS_SERVICE_MARKER, DEPLOYMENT_TYPE=MODEL}, macAddress=null, hostConfig=null, stopSignal=null, healthcheck=null, networkingConfig=null}
[INFO ] c.s.d.c.DefaultDockerClient c.s.d.c.DefaultDockerClient.createContainer.635 Creating container with ContainerConfig: ContainerConfig{hostname=null, domainname=null, user=null, attachStdin=null, attachStdout=null, attachStderr=null, portSpecs=null, exposedPorts=[9091], tty=null, openStdin=null, stdinOnce=null, env=[SIDECAR_PORT=8081, SERVICE_ID=1, MODEL_DIR=/model, SIDECAR_HOST=192.168.90.61, APP_PORT=9091], cmd=null, image=hydrosphere/serving-grpc-runtime-spark-2_1:0.0.1, volumes={}, workingDir=null, entrypoint=null, networkDisabled=null, onBuild=null, labels={SERVICE_ID=1, RUNTIME_ID=2, SERVICE_NAME=r2m1e0, HS_SERVICE_MARKER=HS_SERVICE_MARKER, DEPLOYMENT_TYPE=APP}, macAddress=null, hostConfig=HostConfig{binds=null, blkioWeight=null, blkioWeightDevice=null, blkioDeviceReadBps=null, blkioDeviceWriteBps=null, blkioDeviceReadIOps=null, blkioDeviceWriteIOps=null, containerIdFile=null, lxcConf=null, privileged=null, portBindings={9091=[PortBinding{hostIp=0.0.0.0, hostPort=}]}, links=null, publishAllPorts=null, dns=null, dnsOptions=null, dnsSearch=null, extraHosts=null, volumesFrom=[s1modelword2vec], capAdd=null, capDrop=null, networkMode=null, securityOpt=null, devices=null, memory=null, memorySwap=null, memorySwappiness=null, memoryReservation=null, nanoCpus=null, cpuPeriod=null, cpuShares=null, cpusetCpus=null, cpusetMems=null, cpuQuota=null, cgroupParent=null, restartPolicy=null, logConfig=null, ipcMode=null, ulimits=null, pidMode=null, shmSize=null, oomKillDisable=null, oomScoreAdj=null, autoRemove=null, pidsLimit=null, tmpfs=null, readonlyRootfs=null, storageOpt=null}, stopSignal=null, healthcheck=null, networkingConfig=null}
[INFO ] c.s.d.c.DefaultDockerClient c.s.d.c.DefaultDockerClient.startContainer.657 Starting container with Id: 853ad5441d894760c17d2aedea93a011c97da1c93d4a5033afd643bd2df2272b
[ERROR] i.h.s.m.ManagerHttpApi i.h.s.m.ManagerHttpApi$$anonfun$1.applyOrElse.74 empty.head
java.lang.UnsupportedOperationException: empty.head
	at scala.collection.immutable.Vector.head(Vector.scala:193) ~[scala-library-2.11.11.jar:?]
	at io.hydrosphere.serving.manager.service.ApplicationManagementServiceImpl$$anonfun$inferAppContract$1.apply(ApplicationManagementService.scala:306) ~[classes/:?]
	at io.hydrosphere.serving.manager.service.ApplicationManagementServiceImpl$$anonfun$inferAppContract$1.apply(ApplicationManagementService.scala:305) ~[classes/:?]
	at scala.util.Success$$anonfun$map$1.apply(Try.scala:237) ~[scala-library-2.11.11.jar:?]
	at scala.util.Try$.apply(Try.scala:192) ~[scala-library-2.11.11.jar:?]
	at scala.util.Success.map(Try.scala:237) ~[scala-library-2.11.11.jar:?]
	at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:237) ~[scala-library-2.11.11.jar:?]
	at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:237) ~[scala-library-2.11.11.jar:?]
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:36) [scala-library-2.11.11.jar:?]
	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55) [akka-actor_2.11-2.5.8.jar:?]
	at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply$mcV$sp(BatchingExecutor.scala:91) [akka-actor_2.11-2.5.8.jar:?]
	at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply(BatchingExecutor.scala:91) [akka-actor_2.11-2.5.8.jar:?]
	at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply(BatchingExecutor.scala:91) [akka-actor_2.11-2.5.8.jar:?]
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:72) [scala-library-2.11.11.jar:?]
	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:90) [akka-actor_2.11-2.5.8.jar:?]
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40) [akka-actor_2.11-2.5.8.jar:?]
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:43) [akka-actor_2.11-2.5.8.jar:?]
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260) [akka-actor_2.11-2.5.8.jar:?]
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339) [akka-actor_2.11-2.5.8.jar:?]
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979) [akka-actor_2.11-2.5.8.jar:?]
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107) [akka-actor_2.11-2.5.8.jar:?]
```

When I try to recreate the same application:
```
[ERROR] i.h.s.m.ManagerHttpApi i.h.s.m.ManagerHttpApi$$anonfun$1.applyOrElse.74 ERROR: duplicate key value violates unique constraint "service_service_name_key"
  Detail: Key (service_name)=(r2m1e0) already exists.
org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "service_service_name_key"
  Detail: Key (service_name)=(r2m1e0) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2477) ~[postgresql-42.1.4.jar:42.1.4]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2190) ~[postgresql-42.1.4.jar:42.1.4]
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:300) ~[postgresql-42.1.4.jar:42.1.4]
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:428) ~[postgresql-42.1.4.jar:42.1.4]
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:354) ~[postgresql-42.1.4.jar:42.1.4]
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:169) ~[postgresql-42.1.4.jar:42.1.4]
	at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:136) ~[postgresql-42.1.4.jar:42.1.4]
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61) ~[HikariCP-2.6.3.jar:?]
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java) ~[HikariCP-2.6.3.jar:?]
	at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$SingleInsertAction$$anonfun$run$8.apply(JdbcActionComponent.scala:509) ~[slick_2.11-3.2.1.jar:?]
	at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$SingleInsertAction$$anonfun$run$8.apply(JdbcActionComponent.scala:506) ~[slick_2.11-3.2.1.jar:?]
	at slick.jdbc.JdbcBackend$SessionDef$class.withPreparedInsertStatement(JdbcBackend.scala:378) ~[slick_2.11-3.2.1.jar:?]
	at slick.jdbc.JdbcBackend$BaseSession.withPreparedInsertStatement(JdbcBackend.scala:433) ~[slick_2.11-3.2.1.jar:?]
	at slick.jdbc.JdbcActionComponent$ReturningInsertActionComposerImpl.preparedInsert(JdbcActionComponent.scala:638) ~[slick_2.11-3.2.1.jar:?]
	at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$SingleInsertAction.run(JdbcActionComponent.scala:506) ~[slick_2.11-3.2.1.jar:?]
	at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:29) ~[slick_2.11-3.2.1.jar:?]
	at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:26) ~[slick_2.11-3.2.1.jar:?]
	at slick.basic.BasicBackend$DatabaseDef$$anon$2.liftedTree1$1(BasicBackend.scala:242) ~[slick_2.11-3.2.1.jar:?]
	at slick.basic.BasicBackend$DatabaseDef$$anon$2.run(BasicBackend.scala:242) ~[slick_2.11-3.2.1.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[?:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[?:1.8.0_131]
	at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_131]
```

--- a/manager/src/it/scala/io/hydrosphere/serving/manager/service/ApplicationServiceITSpec.scala
+++ b/manager/src/it/scala/io/hydrosphere/serving/manager/service/ApplicationServiceITSpec.scala
@@ -1,0 +1,56 @@
+package io.hydrosphere.serving.manager.service
+
+import akka.testkit.TestProbe
+import io.hydrosphere.serving.manager.model._
+import io.hydrosphere.serving.manager.service.actors.RepositoryIndexActor
+import io.hydrosphere.serving.manager.test.FullIntegrationSpec
+import org.scalatest.BeforeAndAfterAll
+
+import scala.concurrent.duration._
+
+class ApplicationServiceITSpec extends FullIntegrationSpec with BeforeAndAfterAll{
+
+  "Application service" should {
+    "create a simple application" in {
+      for {
+        version <- managerServices.modelManagementService.buildModel(1, None)
+        appRequest = ApplicationCreateOrUpdateRequest(
+          id = None,
+          name = "testapp",
+          executionGraph = ApplicationExecutionGraph(
+            stages = List(
+              ApplicationStage(
+                services = List(
+                  ServiceWeight(
+                    serviceDescription = ServiceKeyDescription(
+                      runtimeId = 1, // dummy runtime id
+                      modelVersionId = Some(version.id),
+                      environmentId = None
+                    ),
+                    weight = 100
+                  )
+                ),
+                signatureName = "default"
+              )
+            )
+          )
+        )
+        app <- managerServices.applicationManagementService.createApplication(appRequest)
+      } yield {
+        assert(app.name === appRequest.name)
+        assert(app.contract === version.modelContract)
+        assert(app.executionGraph === appRequest.executionGraph)
+      }
+    }
+  }
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    val indexProbe = TestProbe()
+    system.eventStream.subscribe(indexProbe.ref, classOf[RepositoryIndexActor.IndexFinished])
+    managerServices.sourceManagementService.addSource(
+      ModelSourceConfig(1, "itsource", LocalSourceParams(getClass.getResource("/models").getPath)).toAux
+    )
+    indexProbe.expectMsg(15.seconds, RepositoryIndexActor.IndexFinished("dummy_model", "itsource"))
+  }
+}

--- a/manager/src/it/scala/io/hydrosphere/serving/manager/service/ApplicationServiceITSpec.scala
+++ b/manager/src/it/scala/io/hydrosphere/serving/manager/service/ApplicationServiceITSpec.scala
@@ -37,6 +37,7 @@ class ApplicationServiceITSpec extends FullIntegrationSpec with BeforeAndAfterAl
         )
         app <- managerServices.applicationManagementService.createApplication(appRequest)
       } yield {
+        println(app)
         assert(app.name === appRequest.name)
         assert(app.contract === version.modelContract)
         assert(app.executionGraph === appRequest.executionGraph)

--- a/manager/src/it/scala/io/hydrosphere/serving/manager/service/ApplicationServiceITSpec.scala
+++ b/manager/src/it/scala/io/hydrosphere/serving/manager/service/ApplicationServiceITSpec.scala
@@ -47,6 +47,7 @@ class ApplicationServiceITSpec extends FullIntegrationSpec with BeforeAndAfterAl
 
   override def beforeAll(): Unit = {
     super.beforeAll()
+    dockerClient.pull("hydrosphere/serving-runtime-dummy:latest")
     val indexProbe = TestProbe()
     system.eventStream.subscribe(indexProbe.ref, classOf[RepositoryIndexActor.IndexFinished])
     managerServices.sourceManagementService.addSource(

--- a/manager/src/main/scala/io/hydrosphere/serving/manager/controller/ApplicationController.scala
+++ b/manager/src/main/scala/io/hydrosphere/serving/manager/controller/ApplicationController.scala
@@ -47,7 +47,7 @@ class ApplicationController(
     post {
       entity(as[ApplicationCreateOrUpdateRequest]) { r =>
         complete(
-          applicationManagementService.createApplications(r)
+          applicationManagementService.createApplication(r)
         )
       }
     }
@@ -67,7 +67,7 @@ class ApplicationController(
     put {
       entity(as[ApplicationCreateOrUpdateRequest]) { r =>
         complete(
-          applicationManagementService.updateApplications(r)
+          applicationManagementService.updateApplication(r)
         )
       }
     }

--- a/manager/src/main/scala/io/hydrosphere/serving/manager/service/ApplicationManagementService.scala
+++ b/manager/src/main/scala/io/hydrosphere/serving/manager/service/ApplicationManagementService.scala
@@ -45,11 +45,11 @@ trait ApplicationManagementService {
 
   def checkApplicationSchema(req: ApplicationCreateOrUpdateRequest): Future[Boolean]
 
-  def createApplications(req: ApplicationCreateOrUpdateRequest): Future[Application]
+  def createApplication(req: ApplicationCreateOrUpdateRequest): Future[Application]
 
   def deleteApplication(id: Long): Future[Unit]
 
-  def updateApplications(req: ApplicationCreateOrUpdateRequest): Future[Application]
+  def updateApplication(req: ApplicationCreateOrUpdateRequest): Future[Application]
 }
 
 class ApplicationManagementServiceImpl(
@@ -191,7 +191,7 @@ class ApplicationManagementServiceImpl(
     )
   }
 
-  def createApplications(req: ApplicationCreateOrUpdateRequest): Future[Application] = executeWithSync {
+  def createApplication(req: ApplicationCreateOrUpdateRequest): Future[Application] = executeWithSync {
     val keys = for {
       stage <- req.executionGraph.stages
       service <- stage.services
@@ -203,7 +203,7 @@ class ApplicationManagementServiceImpl(
     for {
       services <- serviceManagementService.fetchServicesUnsync(keySet)
       existedServices = services.map(_.toServiceKeyDescription)
-      _ <- startServices(keySet -- existedServices)
+      s <- startServices(keySet -- existedServices)
       inferredApp <- reqToApplication(req)
       createdApp <- applicationRepository.create(inferredApp)
     } yield {
@@ -227,7 +227,7 @@ class ApplicationManagementServiceImpl(
       }
     }
 
-  def updateApplications(req: ApplicationCreateOrUpdateRequest): Future[Application] =
+  def updateApplication(req: ApplicationCreateOrUpdateRequest): Future[Application] =
     executeWithSync {
       applicationRepository.get(req.id.getOrElse(throw new IllegalArgumentException(s"id required $req")))
         .flatMap {
@@ -262,11 +262,12 @@ class ApplicationManagementServiceImpl(
     }
   }
 
-  private def startServices(keysSet: Set[ServiceKeyDescription]): Future[Seq[Service]] =
-    serviceManagementService.fetchServicesUnsync(keysSet).flatMap(services => {
+  private def startServices(keysSet: Set[ServiceKeyDescription]): Future[Seq[Service]] = {
+    logger.debug(keysSet)
+    serviceManagementService.fetchServicesUnsync(keysSet).flatMap { services =>
       val toAdd = keysSet -- services.map(_.toServiceKeyDescription)
 
-      Future.traverse(toAdd.toSeq)(key => {
+      Future.traverse(toAdd.toSeq) { key =>
         serviceManagementService.addService(CreateServiceRequest(
           serviceName = key.toServiceName(),
           runtimeId = key.runtimeId,
@@ -274,8 +275,9 @@ class ApplicationManagementServiceImpl(
           environmentId = key.environmentId,
           modelVersionId = key.modelVersionId
         ))
-      })
-    })
+      }
+    }
+  }
 
   private def removeServiceIfNeeded(keysSet: Set[ServiceKeyDescription], applicationId: Long): Future[Unit] =
     applicationRepository.getKeysNotInApplication(keysSet, applicationId)
@@ -299,6 +301,7 @@ class ApplicationManagementServiceImpl(
   }
 
   private def inferAppContract(application: ApplicationCreateOrUpdateRequest): Future[ModelContract] = {
+    logger.debug(application)
     application.executionGraph.stages match {
       case stage :: Nil if stage.services.lengthCompare(1) == 0 => // single model version
         val serviceDesc = stage.services.head

--- a/manager/src/main/scala/io/hydrosphere/serving/manager/service/ServiceManagementService.scala
+++ b/manager/src/main/scala/io/hydrosphere/serving/manager/service/ServiceManagementService.scala
@@ -153,46 +153,46 @@ class ServiceManagementServiceImpl(
       })
     })
 
-  private def fetchModel(modelId: Option[Long]): Future[Option[ModelVersion]] =
+  private def fetchModel(modelId: Option[Long]): Future[Option[ModelVersion]] = {
+    logger.debug(modelId)
     modelId match {
       case Some(x) =>
         modelVersionRepository.get(x)
           .map(s => s.orElse(throw new IllegalArgumentException(s"Can't find ModelVersion with id=$x")))
       case None => Future.successful(None)
     }
+  }
 
-  private def fetchServingEnvironment(environmentId: Option[Long]): Future[Option[Environment]] = environmentId match {
-    case Some(x) => x match {
-      case AnyEnvironment.`anyEnvironmentId` =>
-        Future.successful(None)
-      case _ =>
-        environmentRepository.get(x)
+  private def fetchServingEnvironment(environmentId: Option[Long]): Future[Option[Environment]] = {
+    logger.debug(environmentId)
+    environmentId match {
+      case Some(x) => x match {
+        case AnyEnvironment.`anyEnvironmentId` =>
+          Future.successful(None)
+        case _ =>
+          environmentRepository.get(x)
+      }
+      case None => Future.successful(None)
     }
-    case None => Future.successful(None)
   }
 
   override def addService(r: CreateServiceRequest): Future[Service] = {
-    //if(r.serviceName) throw new IllegalArgumentException
+    logger.debug(r.toString)
     //TODO ADD validation for names manager,gateway + length + without space and special symbols
-    fetchServingEnvironment(r.environmentId).flatMap(svEnv => {
-      fetchModel(r.modelVersionId).flatMap(modelVersion => {
-        runtimeRepository.get(r.runtimeId).flatMap {
-          case None => throw new IllegalArgumentException(s"Can't find Runtime with id=${r.runtimeId}")
-          case Some(runtime) =>
-            serviceRepository.create(r.toService(runtime, modelVersion, svEnv))
-              .flatMap(newService => {
-                cloudDriverService.deployService(newService).flatMap(cloudService => {
-                  serviceRepository.updateCloudDriveId(cloudService.id, Some(cloudService.cloudDriverId))
-                    .map(_ => {
-                      val s = newService.copy(cloudDriverId = Some(cloudService.cloudDriverId))
-                      internalManagerEventsPublisher.serviceChanged(s)
-                      s
-                    })
-                })
-              })
-        }
-      })
-    })
+    for {
+      svEnv <- fetchServingEnvironment(r.environmentId)
+      modelVersion <- fetchModel(r.modelVersionId)
+      maybeRuntime <- runtimeRepository.get(r.runtimeId)
+      runtime = maybeRuntime.getOrElse(throw new IllegalArgumentException(s"Can't find Runtime with id=${r.runtimeId}"))
+      dService = r.toService(runtime, modelVersion, svEnv)
+      cloudService <- cloudDriverService.deployService(dService)
+      newService <- serviceRepository.create(dService)
+      _ <- serviceRepository.updateCloudDriveId(cloudService.id, Some(cloudService.cloudDriverId))
+    } yield {
+      val s = newService.copy(cloudDriverId = Some(cloudService.cloudDriverId))
+      internalManagerEventsPublisher.serviceChanged(s)
+      s
+    }
   }
 
   override def getService(serviceId: Long): Future[Option[Service]] =

--- a/manager/src/main/scala/io/hydrosphere/serving/manager/service/ServiceManagementService.scala
+++ b/manager/src/main/scala/io/hydrosphere/serving/manager/service/ServiceManagementService.scala
@@ -176,6 +176,26 @@ class ServiceManagementServiceImpl(
     }
   }
 
+  def createAndDeploy(
+    r: CreateServiceRequest,
+    runtime: Runtime,
+    modelVersion: Option[ModelVersion],
+    svEnv: Option[Environment]
+  ): Future[Service] = {
+    val dService = r.toService(runtime, modelVersion, svEnv)
+    serviceRepository.create(dService).flatMap { newService =>
+      val future = cloudDriverService.deployService(newService).flatMap { cloudService =>
+        serviceRepository.updateCloudDriveId(cloudService.id, Some(cloudService.cloudDriverId)).map { _ =>
+          newService.copy(cloudDriverId = Some(cloudService.cloudDriverId))
+        }
+      }
+      future.onFailure {
+        case _ => serviceRepository.delete(dService.id)
+      }
+      future
+    }
+  }
+
   override def addService(r: CreateServiceRequest): Future[Service] = {
     logger.debug(r.toString)
     //TODO ADD validation for names manager,gateway + length + without space and special symbols
@@ -184,14 +204,10 @@ class ServiceManagementServiceImpl(
       modelVersion <- fetchModel(r.modelVersionId)
       maybeRuntime <- runtimeRepository.get(r.runtimeId)
       runtime = maybeRuntime.getOrElse(throw new IllegalArgumentException(s"Can't find Runtime with id=${r.runtimeId}"))
-      dService = r.toService(runtime, modelVersion, svEnv)
-      cloudService <- cloudDriverService.deployService(dService)
-      newService <- serviceRepository.create(dService)
-      _ <- serviceRepository.updateCloudDriveId(cloudService.id, Some(cloudService.cloudDriverId))
+      asd <- createAndDeploy(r, runtime, modelVersion, svEnv)
     } yield {
-      val s = newService.copy(cloudDriverId = Some(cloudService.cloudDriverId))
-      internalManagerEventsPublisher.serviceChanged(s)
-      s
+      internalManagerEventsPublisher.serviceChanged(asd)
+      asd
     }
   }
 

--- a/manager/src/main/scala/io/hydrosphere/serving/manager/service/clouddriver/CloudDriverService.scala
+++ b/manager/src/main/scala/io/hydrosphere/serving/manager/service/clouddriver/CloudDriverService.scala
@@ -2,6 +2,7 @@ package io.hydrosphere.serving.manager.service.clouddriver
 
 import io.hydrosphere.serving.manager.model.api.ModelType
 import io.hydrosphere.serving.manager.model.Service
+import org.apache.logging.log4j.scala.Logging
 
 import scala.concurrent.Future
 
@@ -58,7 +59,7 @@ case class CloudService(
   instances: Seq[ServiceInstance]
 )
 
-trait CloudDriverService {
+trait CloudDriverService extends Logging {
 
   val LABEL_SERVICE_ID = "SERVICE_ID"
   val LABEL_SERVICE_NAME = "SERVICE_NAME"
@@ -111,6 +112,7 @@ trait CloudDriverService {
   }
 
   protected def getRuntimeLabels(service: Service): Map[String, String] = {
+
     Map[String, String](
       LABEL_SERVICE_ID -> service.id.toString,
       LABEL_HS_SERVICE_MARKER -> LABEL_HS_SERVICE_MARKER,

--- a/manager/src/main/scala/io/hydrosphere/serving/manager/service/clouddriver/LocalCloudDriverService.scala
+++ b/manager/src/main/scala/io/hydrosphere/serving/manager/service/clouddriver/LocalCloudDriverService.scala
@@ -9,6 +9,7 @@ import io.hydrosphere.serving.manager.ManagerConfiguration
 import io.hydrosphere.serving.manager.model.api.ModelType
 import io.hydrosphere.serving.manager.model.Service
 import io.hydrosphere.serving.manager.service.InternalManagerEventsPublisher
+import org.apache.logging.log4j.scala.Logging
 
 import scala.concurrent.{ExecutionContext, Future}
 import collection.JavaConversions._
@@ -21,7 +22,7 @@ class LocalCloudDriverService(
   dockerClient: DockerClient,
   managerConfiguration: ManagerConfiguration,
   internalManagerEventsPublisher: InternalManagerEventsPublisher
-)(implicit val ex: ExecutionContext) extends CloudDriverService {
+)(implicit val ex: ExecutionContext) extends CloudDriverService with Logging {
 
   override def serviceList(): Future[Seq[CloudService]] = Future({
     postProcessAllServiceList(getAllServices())
@@ -68,11 +69,12 @@ class LocalCloudDriverService(
     publishPorts
   }
 
-  override def deployService(service: Service): Future[CloudService] = Future.apply({
+  override def deployService(service: Service): Future[CloudService] = Future.apply{
+    logger.debug(service)
     val modelContainerId = service.model.map(_ => startModel(service))
-    val javaLabels = mapAsJavaMap(getRuntimeLabels(service) ++ Map(
+    val javaLabels = getRuntimeLabels(service) ++ Map(
       LABEL_SERVICE_NAME -> service.serviceName
-    ))
+    )
 
     val envMap = service.configParams ++ Map(
       ENV_MODEL_DIR -> DEFAULT_MODEL_DIR.toString,
@@ -84,9 +86,9 @@ class LocalCloudDriverService(
 
     val builder = createMainApplicationHostConfigBuilder()
 
-    modelContainerId.foreach(_ => {
+    modelContainerId.foreach{_ =>
       builder.volumesFrom(generateModelContainerName(service))
-    })
+    }
 
     val c = dockerClient.createContainer(ContainerConfig.builder()
       .image(service.runtime.toImageDef)
@@ -100,7 +102,7 @@ class LocalCloudDriverService(
     val cloudService = fetchById(service.id)
     internalManagerEventsPublisher.cloudServiceDetected(Seq(cloudService))
     cloudService
-  })
+  }
 
   private def collectCloudService(containers: Seq[Container]): Seq[CloudService] = {
     val map = containers.groupBy(c => c.labels().get(LABEL_SERVICE_ID))

